### ISSUE TITLE
libmongo-client submodule vs make dist

### ISSUE THREAD

**Issue by algernon** [balabit/syslog-ng#692]
While trying to find a way to automatically - and reliably - create a `make dist` tarball from a git checkout, I ran into the following problem:

```
$ ./configure && make dist
[....]
 (cd modules/afmongodb/libmongo-client && make --no-print-directory top_distdir=../../../syslog-ng-3.7.1 distdir=../../../syslog-ng-3.7.1/modules/afmongodb/libmongo-client \
     am__remove_distdir=: am__skip_length_check=: am__skip_mode_fix=: distdir)
make[2]: *** No rule to make target 'distdir'.  Stop.
Makefile:11839: recipe for target 'distdir' failed
make[1]: *** [distdir] Error 1
Makefile:11938: recipe for target 'dist' failed
make: *** [dist] Error 2
```

It tries to do a `make dist` in `modules/afmongodb/libmongo-client`, but that fails, because I have `libmongo-client-dev` installed, and the system version is preferred (which is a good default), so the embedded copy is neither configured, nor built.

I can work it around by using `--with-libmongo-client=internal` at configure-time, but... that's a tad ugly. Wouldn't it make more sense to nuke the embedded libmongo-client? It is included in recent distros, and for EPEL & the like, it is packaged separately anyway. For the tiny number of people who build syslog-ng from source and need the mongodb destination, the extra step of separately downloading the library should not pose any challenge.

I'd do the same with the embedded `json-c`, `ivykis`, and consider testing whether it can be done for `librabbitmq-c` as well (but right now, that does not support building with a system-wide lib, which may worth checking, if that's still the case).

In short, the embedded copies are increasingly less useful (as far as I see), yet, are an increasingly big burden. Are they really worth it?

**Comment by dnsjts** [comment]
What do you think about this patch?
```diff
diff --git a/modules/afmongodb/Makefile.am b/modules/afmongodb/Makefile.am
index f2d1c6e..d4f2546 100644
--- a/modules/afmongodb/Makefile.am
+++ b/modules/afmongodb/Makefile.am
@@ -1,6 +1,5 @@
-DIST_SUBDIRS                                   += modules/afmongodb/libmongo-client
-
 if LIBMONGO_INTERNAL
+DIST_SUBDIRS                                   += modules/afmongodb/libmongo-client
 CLEAN_SUBDIRS                                  += @LIBMONGO_SUBDIRS@
 lmc_EXTRA_DEPS                                 = modules/afmongodb/libmongo-client/src/libmongo-client.la
```

I would test it, but it seems, that I have got some issues with my virtual machines, and maybe this could help you.

**Comment by algernon** [comment]
As Bazsi mentioned in #694, the contents of the dist tarball should not depend on what one has available when running `make dist`. So I'd much prefer dropping the submodules altogether.

**Comment by dnsjts** [comment]
I think nuking the libmongo-client would cause more harm than good.
For example installing the appropiate version of the libmongo-client from the official repositories is problematic on the following distributions: Arch, Gentoo, Ubuntu (before vivid).
Of course I could compile them from source and install them, but I'm afraid of these hybrid solutions (i.e. installing from repositories and installing from source).

The dist tarball would contain exactly what you have asked at configuration time, therefore the original behaviour can be restored, and you can create tarballs with less size.

**Comment by pasztor** [comment]
I think the incorporated submodules worth some design / refactor.
As far as I remember MPlayer had the same problem earlier. (eg. ffmpeg and other libraries which were copied in and tested by hand.)
Maybe that worth some time to look what they did.
I also found an article in the deep of my mailbox about the topic:
http://googletesting.blogspot.hu/2015/05/multi-repository-development.html

These are not "solutions", just suggestions, <IMHO> what we should investigate</IMHO>.

**Comment by bazsi** [comment]
The original idea was to keep them as long as the versions we depend on are
not generally available.

It is a lot different than the mplayer case as we don't intend to modify
these libs or if we do, we do that upstream.

So, if we have popular distros that are lacking the required versions, we
should keep them.

Perhaps a ticket to the respective maintainers would be great, in order to
push distros to the right direction.

Anyone volunteering to send theese tickets? It would be a great
contribution.

**Comment by algernon** [comment]
> For example installing the appropiate version of the libmongo-client from the official repositories is problematic on the following distributions: Arch, Gentoo, Ubuntu (before vivid).

Yet, `riemann-c-client` is not bundled with syslog-ng, even though it is even less generally available than `libmongo-client`. Not bundling it is a great incentive for distributions to pick up the new versions. In case of Arch and Gentoo, that's fairly simple. For Ubuntu, Vivid is the latest LTS anyway. I don't see a problem with the mongodb destination not being available on older releases by default.

There are backports to cover the older releases. I used to build half the world to support Ubuntu Lucid. That's the correct way, in my not so humble opinion: depend on new versions when needed, provide or encourage backport repos if there's a demand (based on a quick call to users, the demand seems to be nonexistent, as far as Debian & Ubuntu is concerned, so, just stop caring about old releases of these OSes).

> As far as I remember MPlayer had the same problem earlier. (eg. ffmpeg and other libraries which were copied in and tested by hand.)

As @bazsi said, that's an entirely different case. MPlayer modified the shipped libraries, syslog-ng doesn't. It shipped with patched ivykis and librabbitmq-c for a while, but not anymore.

> So, if we have popular distros that are lacking the required versions, we
should keep them.

I respectfully disagree. Distro maintainers or other interested parties can backport the required packages if there's a need. On the Debian/Ubuntu side, the demand does not exist. On the RedHat side, backports are already made, as far as I see. For Arch and Gentoo, which are - if I remember correctly - using a single, release-less repository, updating libmongo-client and the rest should be much less painful than for Debian & RedHat-based distros.

**Comment by lbudai** [comment]
Hi,

to be honest the nice and clean solution would be to drop the submodules from syslog-ng.
But I'm not sure that this is the right solution...

What if we would provide two dist tarballs: default is the minimal one (without submodules)
and other is the fat one (with submodules, make dist-fat)?

And after a while (which should be communicated) we would drop the submodules?
(during this time `someone` could try to contact with distros...)

L.

**Comment by czanik** [comment]
Hi,
Difficult question. Fedora enforces that bundled libraries can't be used, I could not update to syslog-ng 3.3 in Fedora until external ivykis worked and was packaged for Fedora.
On the other hand on my other platforms (openSUSE, etc.) this is not enforced at all, and I rather receive complaints that syslog-ng has too many dependencies. Also, it is usually quite difficult to introduce a new package and it gives considerable administrative overhead. I feel this work especially superfluous knowing that these would not be used outside of syslog-ng.
CzP

**Comment by algernon** [comment]
> I feel this work especially superfluous knowing that these would not be used outside of syslog-ng.

libmongo-client is used outside of syslog-ng (by rsyslog), too. syslog-ng never bundled riemann-c-client, even though it is even less likely to be present than libmongo-client.

> to be honest the nice and clean solution would be to drop the submodules from syslog-ng.
But I'm not sure that this is the right solution...

rsyslog does not embed libmongo-client. collectd doesn't embed all the stuff it has optional support for compiling with. Pretty much the only software that does embed copious amounts of 3rd party libs, are those that also patch it.

> What if we would provide two dist tarballs: default is the minimal one (without submodules)
and other is the fat one (with submodules, make dist-fat)?

Mngh. That sounds even uglier than the status quo, and means more work for you. (Distros can already work around most of the submodule problems in one way or  the other.)

How about dropping submodules for 3.8, and if there's a demand for them, revert the change?

> On the other hand on my other platforms (openSUSE, etc.) this is not enforced at all, and I rather receive complaints that syslog-ng has too many dependencies.

Well, that's why you split out `syslog-ng-mod-$foo` packages, or make your monolithic package only recommend optional libraries. 

The `syslog-ng-core` package on Debian has only ~3 more dependencies compared to the old monolithic package: `libuuid1`, `libivykis0`, and some systemd stuff. It also doesn't depend on libdbi, which makes the total only 2 more, both of which will be installed by default on any recent Debian system. So on a default install, syslog-ng-core installs less extra packages than the old monolithic one did. That's quite a progress, I'd say.

Or you can go the other route, where you depend on the libraries absolutely necessary for the default config, and only suggest/recommend the rest. This is harder to set up, but allows you to preserve a monolithic package. (On Debian, collectd does this, for example.)

**Comment by bazsi** [comment]
I am fine with dropping those libraries if the last LTS/stable of Debian/RedHat/Fedora/Ubuntu/SUSE/OpenSUSE/FreeBSD has the requisite version.

It would become more difficult to compile syslog-ng on commercial UNIXes like Solaris, but not by much.

On the other hand eventlog should be bundled into syslog-ng and completely discontinued as a separate package.

**Comment by algernon** [comment]
That plan works for me, too.

**Comment by algernon** [comment]
libmongo-client is not used anymore, and I no longer find this issue relevant, either, so I'm closing it.
